### DISCUSSION TITLE
How to get braintree working in Sandbox?

### DISCUSSION THREAD

**Discussion by kudapara** [pay-rails/pay#315] [General]
Good day. I have been struggling for the past week to get Braintree working.

Stripe was so simple and straight forward and I can test subscriptions and charges easily.
However with Braintree, I have not yet made a single successful test transaction in sandbox.

I have created a test Paypal app and connected it with Braintree and everything seems smooth up until the point where this gem has to charge the customer, that is when I get a `declined` error with no additional context.

Is there anyone who has been able to successfully integrate Braintree?

Here is a screenshot of the error.
<img width="1440" alt="Screenshot 2021-04-22 at 15 35 22" src="https://user-images.githubusercontent.com/13519184/115724730-8147b100-a381-11eb-8bf6-3ccb59df250a.png">

**Comment by excid3** [comment]
There are a couple different ways to test PayPal with Braintree: https://developers.braintreepayments.com/guides/paypal/testing-go-live/ruby

**Comment by kudapara** [comment]
Thank you @excid3. 

I have followed the link, and I have unlinked the paypal account, now using the Mocked flow. I am now getting a different error which is good progress 😅. I am passing the nonce from the browser, but somehow it is saying no customer found. Have you ever got this error before?


<img width="1095" alt="Screenshot 2021-04-22 at 16 32 59" src="https://user-images.githubusercontent.com/13519184/115732756-8f4d0000-a388-11eb-8618-6c0bfb06742c.png">

**Comment by excid3** [comment]
Did you delete that Braintree customer with that ID by chance?

**Comment by kudapara** [comment]
I think so. I had purged the test data to start implementing your proposed solution on a fresh slate. I had not also purged my local reference.

I finally got it working by setting `user.processor = :fake_processor` in the console and then restarting the payment flow. Somehow that cleared the Braintree customer and it worked.

I have a question. The app i'm building is an ecommerce platform and on checkout, do I have to check if the customer has the processor in their user object and hide the Paypal button/stripe fields, so that we can charge them with the payment details on file, so they do not have to go through the process of entering their details and approving our app?

I ask this because I want to know if the card_tokens expire, or they are valid until the customer manually unsubscribes.

**Comment by excid3** [comment]
If you deleted the customer in Braintree, you'll just want to update your user in the database `update(processor_id: nil)` so it matches. 👍 

The `fake_processor` will not contact Braintree or anything, so it won't really verify that your Braintree setup works.

For my apps, I display both the PayPal and Stripe fields and then submit `stripe` or `braintree` as a hidden field from Javascript and assign that before making the charge.

Cards are stored in the payment processor, so they don't expire (as long as you assign it to the customer which Pay does automatically if you set `card_token`).
